Adding documentation for FM and COOP/COEP (#1)
gchatz22/struts-site: source/core-developers/fetch-metadata-interceptor.md

---
layout: default
title: Fetch Metadata Interceptor
parent:
    title: Interceptors
    url: interceptors.html
---

# Fetch Metadata Interceptor

## Description

> An interceptor that implements Fetch Metadata policy on incoming requests used to protect against CSRF, XSSI, and cross-origin information leaks. Uses a deafault Resource Isolation Policy to filter the requests allowed to be processed.

Reviewer: ```suggestion
An interceptor that implements Fetch Metadata on incoming requests used to protect against CSRF, XSSI, and cross-origin information leaks. Uses a default Resource Isolation Policy to programmatically reject cross-origin requests.
```

> A Resource Isolation Policy is a strong defense-in-depth mechanism that prevents the resources on a server from being requested by external websites. This policy can be enabled for all endpoints of the application or the endpoints that are meant to be loaded in a cross-site context can be exempted from applying the policy.

Reviewer: ```suggestion
A Resource Isolation Policy is a strong defense in-depth mechanism that prevents the resources on a server from being requested by external websites. This policy can be enabled either for all endpoints of the application and  endpoints that are meant to be loaded in a cross-site context can be exempted from the policy.
```

> The browser provides information about the context of an HTTP request in a set of Sec-Fetch-* headers. This allows the server processing the request to make decisions on whether the request should be accepted or rejected based on the preferred resource isolation policy. At a very high level, the Fetch Metadata policy rejects requests with

Reviewer: Would be great if we could code-quote Sec-Fetch-*

> The browser provides information about the context of an HTTP request in a set of Sec-Fetch-* headers. This allows the server processing the request to make decisions on whether the request should be accepted or rejected based on the preferred resource isolation policy. At a very high level, the Fetch Metadata policy rejects requests with

Reviewer: I like the codeblock below but chances are it will become obsolete with tiny modifications to the default policy!

```suggestion
The browser provides information about the context of an HTTP request in a set of Sec-Fetch-* headers. This allows the server processing the request to make decisions on whether the request should be accepted or rejected based on the preferred resource isolation policy. Struts provides a default Resource Isolation Policy that rejects cross-origin requests that aren't top level navigations.
```

```
Sec-Fetch-Site == 'cross-site' AND (Sec-Fetch-Mode != 'navigate'/'nested-navigate' OR method NOT IN [GET, HEAD])
```

> Refer to [Implementing a Resource Isolation Policy](https://web.dev/fetch-metadata/#implementing-a-resource-isolation-policy) for further information on effectively implementing Fetch Metadata.

Reviewer: ```suggestion
Refer to [Implementing a Resource Isolation Policy](https://web.dev/fetch-metadata/#implementing-a-resource-isolation-policy) for further information on implementing effective Resource Isolation Policies.
```

Fetch Metadata is supported in all major browsers

## Parameters

- `exemptedPaths` - Set of opt out endpoints that are meant to serve cross-site traffic
- `resourceIsolationPolicy` - Instance of Resource Isolation Policy class implementing the logic for the requests filtering
- `VARY_HEADER_VALUE` - static vary header value for each vary response headers

> - `SC_FORBIDDEN` - static string 403 status code returned every time a request is rejected by Fetch Metadata

Reviewer: I believe these are not parameters that users are allowed to configure, right? there is no `<param name='vary_header_value'>cheese</param>`. If that's the case, we shouldn't list them here

Reviewer 2: Yes same for SC_FORBIDDEN. Wasn't sure how much in detail I should go to
